core/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java

```diff
@@ -194,7 +200,9 @@ public static State fromString(String state) {
     private final Version indexUpgradedVersion;
     private final org.apache.lucene.util.Version minimumCompatibleLuceneVersion;
 
-    private IndexMetaData(String index, long version, State state, Settings settings, ImmutableOpenMap<String, MappingMetaData> mappings, ImmutableOpenMap<String, AliasMetaData> aliases, ImmutableOpenMap<String, Custom> customs) {
+    private IndexMetaData(String index, long version, State state, Settings settings, ImmutableOpenMap<String, MappingMetaData> mappings,
+                          ImmutableOpenMap<String, AliasMetaData> aliases, ImmutableOpenMap<String, Custom> customs,
+                          Set<String> primaryCandidateAllocationIds) {
```

Comment:
also we need those *PER* shard.. so the allocation ids for shard 0, alloc ids for shard 1 etc.